MultipeerConnectivityでデバイス間接続を可能にする

tsuu32 opened the pull request:
## 概要

MultipeerConnectivityのラッパーライブラリConnectionController.swiftを作成。
ConnectionController.swiftで初期化、advertising、browsingが可能。
ConnectionControllerDelegate.swiftのメソッドを実装することで、browsing中に見つかった時の処理、データが送られてきた時の処理をかける。

## やったこと

- [x] 初期表示でのWelcomeViewControllerのDoneボタンをdisable
  - [ ] 初期表示でのWelcomeViewControllerのDoneボタンを隠す
- [x] Join as the DJボタンでセッションを初期化してadvertising開始
  - [x] 再生中の曲のインデックスの自己管理
- [x] Join as a ListenerボタンでDJ選択画面の表示 (Browsing開始)
  - [x] DJ選択画面で接続可能なDJをTableViewで表示
  - [x] DJ選択画面で表示された接続可能なDJを表すTableViewCellをタップして接続
- [x] connectionController(didReceiveData data: Data, from peerID: MCPeerID)に受け取った時の処理をかけるようにする
- [ ] DJになった後に自分がDJであることをわかる表示 (ナビゲーションバーに色付け？)

#### 重要な情報

Notificationは使っていません。

デバイス間で接続するのはとりあえずできました。
送受信に関してですが、送る方は @bldsky が頑張ってます。
受け取る処理がバグってます。
(WelcomeViewControllerでのConnectionController.shared.delegate = selfがバグってるせい？)

## やってほしいこと

「リスナは何を送ればいいか」、「DJは受け取った時何をするか」を知りたい。

それと、DJが持つReuestQueue(音楽再生のキュー)の情報を、リスナにどう送るかなどを考えなくちゃならない。(キューのリスナとの同期、リスナはキューから削除できるのか、スキップできるのかなど)

tsuu32 committed:
Join as the DJのViewを削除

Main.storyboard:
WelcomeViewController.swift:
Join as the DJのViewを削除した。
[Join as the DJ]ボタンをタップすると、joinAsDJ関数が呼ばれる（Viewをdismissする）。

tsuu32 committed:
Join as a ListenerのViewでTable Viewを作成

Main.storyboard:
TableViewとTableViewCellを追加（Constraintsもしっかり設定）

ListenerConnectionViewController.swift:
作成
tableView.dataSource = selfを追加

ListenerConnectableDJsTableViewCell:
作成

tsuu32 committed:
デバイス間の接続ができるようになった

ConnectionController.swift
追加
ConnectionControllerDelegate.swift:
追加
WelcomeViewController.swift:
ConnectionControllerを用いるようにした
ListenerConnectionViewController.swift
ConnectionControllerを用いるようにした

bldsky committed:
つなげてリクエストを投げられるように

amylaseF85 commented on DJYusaku/ConnectionController.swift:
```diff
@@ -0,0 +1,131 @@
+import Foundation
+import MultipeerConnectivity
+
+class ConnectionController: NSObject {
+    static let shared = ConnectionController()
+    
+    public weak var delegate: ConnectionControllerDelegate?
+    
+    let serviceType = "djyusaku"
+    
+    var peerID :MCPeerID!
+    var session: MCSession!
+    var advertiser: MCNearbyServiceAdvertiser!
+    var browser: MCNearbyServiceBrowser!
+    
+    var connectableDJs: [MCPeerID] = []
+    var connectedDJ: MCPeerID!
+    
+    var isParent: Bool!
+    
+    func initialize(isParent: Bool, displayName: String) {
+        self.isParent = isParent
+        self.connectableDJs = []
+        
+        self.peerID = MCPeerID(displayName: displayName)
+        self.session = MCSession(peer: self.peerID)
+        session.delegate = self
+        
+        advertiser = MCNearbyServiceAdvertiser(peer: self.peerID, discoveryInfo: nil, serviceType: self.serviceType)
+        advertiser.delegate = self
+
+        browser = MCNearbyServiceBrowser(peer: self.peerID, serviceType: self.serviceType)
+        browser.delegate = self
```
```swift
  var connectableDJs: [MCPeerID] = []
```
と
```swift
self.connectableDJs = []
```
どっちかが意味ないと思う

tsuu32 replied:
`func initialize(isParent: Bool, displayName: String)`の中のほうは絶対必要なので、修正するなら
```swift
  var connectableDJs: [MCPeerID] = []
```
のほうでしょう。
意味はないのかもしれませんが、これのせいでバグってるとかはないので、修正しなくてもいい気はします。

amylaseF85 commented on DJYusaku/ListenerConnectionViewController.swift:
```diff
@@ -0,0 +1,84 @@
+    }
+    
+
+    /*
+    // MARK: - Navigation
+
+    // In a storyboard-based application, you will often want to do a little preparation before navigation
+    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
+        // Get the new view controller using segue.destination.
+        // Pass the selected object to the new view controller.
+    }
+    */
```
初期状態からあるように見えるんですがコレはこの後必要になる？

tsuu32 replied:
これはXcodeでCocoa Touch Classなファイルを作るときに自動で挿入されるコメントアウトです。

今後必要になることはないと思われます。

amylaseF85 replied:
問題なさそうなら消しておいて下さい

amylaseF85 reviewed (commented):
デバックとかやってないんでアレだけど、ぱっと見で気になったところを指摘しておきます

yaplus commented on DJYusaku/ConnectionControllerDelegate.swift:
```diff
@@ -0,0 +1,18 @@
+
+    // 接続可能なピアが見つかったとき
+    func connectionController(connectableDevicesChanged devices: [MCPeerID])
+}
```
didReceiveDataとconnectableDevicesChangedでdid＋現在形と過去形になってるの若干気になる

tsuu32 replied:
ab032c3 でconnectableDevicesChangedをdidChangeConnectableDevicesに改名しました

yaplus commented on DJYusaku/ConnectionController.swift:
```diff
@@ -0,0 +1,131 @@
+        // TODO: bldsky が送られてき時の動作を書く
+        DispatchQueue.main.async {
+            self.delegate?.connectionController(didReceiveData: data, from: peerID)
+        }
```
Delegateの処理をメインスレッドで実行するのをやめなさーい😾

tsuu32 replied:
f49e719 で修正しました。

yaplus reviewed (changes requested):
Delegate呼び出しがメインスレッドになってるところだけ直したらOKです

bldsky committed:
completionを記述

bldsky committed:
リスナーのsendの送り先をDJに限定するよう変更

tsuu32 committed:
Merge pull request #2 from bldsky/feature/send-p2p

Feature/send p2p

tsuu32 committed:
connectableDevicesChangedをdidChangeConnectableDevicesにリネーム

ConnectionController.swift:
ConnectionControllerDelegate.swift:
ListenerConnectionViewController.swift:
didChangeConnectableDevicesを使う

yaplus commented:
Notification以外は過去分詞の方に揃えた方がいい気がしたんだけど、どうだろう（didReceiveData:data とかreceivedData: dataの方が良くない？）

tsuu32 commented:
https://developer.apple.com/documentation/multipeerconnectivity/mcsessiondelegate
MCSessionDelegateのメソッドらはdid+動詞が使われてるよ

tsuu32 commented:
@bldsky 
今気づいたんだけど、MCSession+sendRequest.swiftファイルがDJYusakuディレクトリの中じゃなくてリポジトリのルートにあるのはミス？それとも意図的？

yaplus commented:
すいません、Coding Guidelines for Cocoaの[Delgate Method](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/CodingGuidelines/Articles/NamingMethods.html#//apple_ref/doc/uid/20001282-1004202-BCIGGFCC)の項読んでたら

>Use “did” or “will” for methods that are invoked to notify the delegate that something has happened or is about to happen.

> Although you can use “did” or “will” for methods that are invoked to ask the delegate to do something on behalf of another object, “should” is preferred.

ってあるしこれで大丈夫っぽいです。

yaplus commented:
MCSession+sendRequest.swiftの位置が直り次第マージします

yaplus commented:
@tsuu32 
多分ミスだと思うからそっちで直しておいてくれると助かる

tsuu32 committed:
MCSession+sendRequest.swiftの位置を変更

MCSession+sendRequest.swift:
リポジトリのルートから、リポジトリのルート/DJYusakuに移動

yaplus reviewed (approved):
みました  
通信するためにバンドル名揃えなきゃいけないのでこれマージして僕のバンドルIDに書き換えたらTestFlight登録します #26

tsuu32 commented:
b10f8fc でMCSession+sendRequest.swiftの位置を変更しました
ビルドして試してもエラーは出なかったので大丈夫だと思います
